### DISCUSSION TITLE
About PS tab autocompletion

### DISCUSSION THREAD

**Discussion by Drllap** [gerardog/gsudo#258] [Ideas]
## Description
Add tab autocompletion. Some CLI applications support tab autocompletion in powershell, like chocolatey. I can do `choco out<tab>` it will autocomplete to `choco outdated`, but I cont to `sudo choco out<tab>`

## Proposed technical details
I'm not event sure this is possible.

**Comment by gerardog** [comment]
Choco auto-complete is not working for me.

I've found [Register-ArgumentCompleter](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/register-argumentcompleter) but seems to works just for argument values, not argument names.

If anyone has an idea on how to make a useful autocomplete, please jump in.

**Reply by Drllap** [reply]
To enable auto complete for choco you need:
`Import-Module $env:ChocolateyInstall\helpers\chocolateyProfile.psm1`

**Comment by gerardog** [comment]
I've asked Chatgpt and provided this interesting approach:

```
$gsudoOptions = @{
    'config' = @('--global', 'loglevel', 'cachemode');
    '--global' = @('loglevel', 'cachemode');
    'loglevel' = @('All', 'Debug', 'Info', 'Warning', 'Error', 'None', '--reset');
    'cachemode' = @('Auto', 'Disabled', 'Explicit', '--reset');
    'cache' = @('on', 'off');
    'status' = @{};
    '-k' = @{};
    '-d' = @('--direct', '--debug', '--copyns', '--copyev');
    '--direct' = @{};
    '--debug' = @{};
    '--copyns' = @{};
    '--copyev' = @{};
    '-n' = @('--new');
    '--new' = @{};
    '-w' = @('--wait');
    '--wait' = @{};
    '--keepShellOpen' = @{};
    '--keepWindow' = @{};
    '--close' = @{};
    '-i' = @('--integrity');
    '--integrity' = @('Untrusted', 'Low', 'Medium', 'MediumPlus', 'High', 'System');
    '-u' = @('--user');
    '--user' = @('username');
    '!!' = @{};
}

Register-ArgumentCompleter -Native -CommandName 'gsudo' -ScriptBlock {
    param($commandName, $wordToComplete, $cursorPosition)
        $lastWord = $wordToComplete.Split(' ')[-1]
        $options = $gsudoOptions[$lastWord]
        if ($options) {
            $options | ForEach-Object {
                New-Object System.Management.Automation.CompletionResult $_
            }
        }
}
```
This may be doable after all!

**Comment by gerardog** [comment]
This has been released in v2.1.0. Have you tried it @Drllap ?

**Reply by Drllap** [reply]
just tested it and now it doesn't appear to be working

**Reply by gerardog** [reply]
Did you ran `Import-Module 'c:\Program Files\gsudo\Current\gsudoModule.psd1'` ?

**Reply by Drllap** [reply]
yes, I have it in my profile:
![image](https://github.com/gerardog/gsudo/assets/3659986/dc0a79b2-82c0-4204-8fac-f9200b44432d)

**Reply by Drllap** [reply]
well, it's actually in `C:\tools\gsudo` for me. I installed with `chocolatey`
![image](https://github.com/gerardog/gsudo/assets/3659986/425d5152-26cb-40d2-b740-d11891165a7a)

**Reply by gerardog** [reply]
If you update to gsudo v2.3.0, you can use `Import-Module gsudoModule` with no path, works with winget/scoop & chocolatey.

**Comment by gerardog** [comment]
Have you updated to gsudo v2.1.0 and only then restarted your powershells / loaded gsudoModule?
![image](https://github.com/gerardog/gsudo/assets/3901474/a48be221-3cec-4709-bcfb-8eb254d16021)

writing `gsudo` then space and tab key should autocomplete with your last command...
![image](https://github.com/gerardog/gsudo/assets/3901474/980a199c-f944-436c-b89a-4292c2ce2734)

**Reply by Drllap** [reply]
Yes I have updated
![image](https://github.com/gerardog/gsudo/assets/3659986/da78ec76-df17-4481-a09f-59835d7c9f06)
And restarted.
If I to as you describe:
![image](https://github.com/gerardog/gsudo/assets/3659986/4b3ba153-a71a-4f72-8ba1-af4cf9a82cc3)

**Reply by gerardog** [reply]
Well, I see the auto complete suggestions right there ( `{echo something}`, `status`, `cache`, `config`, etc).Those are being suggested by 
 `gsudo` Module. But looks like you also installed something that is making the suggestions show below, instead of inline... probably the arrow keys let you choose one of the suggestions...

**Reply by Drllap** [reply]
Yes but the `{echo something}` is just coming from history. Things like `sudo choco out<tab>` or `sudo git st` don't trigger the tab autocompletion for `choco`/`git`

**Reply by gerardog** [reply]
They are not coming from history automatically. It is the gsudo auto complete explicitly suggesting your last 3 commands....

If you start your line with sudo/gsudo, only the gsudo auto complete will kick in. But if you use gsudo [Powershell syntax](https://github.com/gerardog/gsudo#usage-from-powershell--powershell-core) that won't be a problem, as soon as you open a  { brace } it will auto-complete other powershell commands. Unless they are native (non-powershell), such as gsudo or choco. You may need to create a ticket for powershell's team to implement that. 

https://github.com/gerardog/gsudo/assets/3901474/739dd357-4226-4e56-b717-a1ecc3e6a263
